src/GravitationalClustering.js

```diff
@@ -0,0 +1,189 @@
+        this.gamma = options.gamma ? options.gamma : 0.2;
+        this.dist = options.dist ? options.dist : Distance.euclidean;
+
+        this.particles = [];
+        this.outliers = [];
+
+        if (X !== undefined && X.length !== 0) {
+            this.set(X, masses);
+        }
+    }
+
+    /**
+     * Set the points and corresponding masses for the algorithm
+     * @param X {Array} - Points to create the clusters.
+     * @param masses {Array} - Mass of each point (Optional).
+     */
+    set(X, masses) {
+        if (X.length === 0) {
+            throw new RangeError('The matrix length should be greater than 0.');
+        }
+        if (masses === undefined) {
+            masses = new Array(X.length);
+        }
+
+        this.particles = [];
+        this.outliers = [];
+
+        var dim = X[0].length;
+        var elements = X.length;
+        var minValues = new Array(dim);
+        var maxValues = new Array(dim);
+
+        for (var i = 0; i < dim; ++i) {
+            minValues[i] = Number.MAX_VALUE;
+            maxValues[i] = Number.MIN_VALUE;
+        }
+
+        this.uf = new UnionFind();
+        this.disjointElems = new Array(elements);
+        for (i = 0; i < elements; ++i) {
+            this.disjointElems[i] = this.uf.add(i);
+            var elem = X[i];
+            var mass = masses[i];
+            if (dim !== elem.length) throw new RangeError('The element at position ' + i + 'must have a size of '
+                                                        + dim + ' but has: ' + elem.length);
+
+            for (var j = 0; j < dim; ++j) {
+                var comp = elem[j];
+                minValues[j] = Math.min(minValues[j], comp);
+                maxValues[j] = Math.max(maxValues[j], comp);
+            }
+
+            this.particles.push(new Particle(elem, mass));
+        }
+
+        this.eps = this.gamma * this.dist(maxValues, minValues);
+    }
+
+    /**
+     * Move two particles closer to each other based on the gravitational constant and the selected distance function.
+     * @param particleA {Particle}
+     * @param particleB {Particle}
+     */
+    move(particleA, particleB) {
+        var deltaA = Particle.moveParticle(particleA, particleB, this.GC, this.dist);
+        for (var i = 0; i < deltaA.length; ++i) {
+            particleA.position[i] += deltaA[i];
+            particleB.position[i] -= deltaA[i];
+        }
+    }
+
+    /**
+     * Merge two particle indices in the X array to the same cluster.
+     * @param a {DisjointSetNode}
+     * @param b {DisjointSetNode}
+     */
+    merge(a, b) {
+        this.uf.union(a, b);
+    }
+
+    /**
+     * Run the algorithm n-iterations partially and return the obtained clusters and outliers.
+     * you can call again this method at the current state.
+     * @param iterations {Number} - Partial iterations.
+     * @return {{outliers: Array, X: Array, y: Array, clusters: Number}}
+     */
+    run(iterations) {
```

Comment:
This doesn't follows the [predictor API](https://github.com/mljs/ml/blob/master/CONTRIBUTING.md#predictors), may be call this function `predict`?